std/permutation/ripemd160/ripemd160block.go

```diff
@@ -0,0 +1,301 @@
+package ripemd160
+
+import (
+	"github.com/consensys/gnark/std/math/uints"
+)
+
+// work buffer indices and roll amounts for one line
+var _n = [80]uint{
+	0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
+	7, 4, 13, 1, 10, 6, 15, 3, 12, 0, 9, 5, 2, 14, 11, 8,
+	3, 10, 14, 4, 9, 15, 8, 1, 2, 7, 0, 6, 13, 11, 5, 12,
+	1, 9, 11, 10, 0, 8, 12, 4, 13, 3, 7, 15, 14, 5, 6, 2,
+	4, 0, 5, 9, 7, 12, 2, 10, 14, 1, 3, 8, 11, 6, 15, 13,
+}
+
+var _r = [80]uint{
+	11, 14, 15, 12, 5, 8, 7, 9, 11, 13, 14, 15, 6, 7, 9, 8,
+	7, 6, 8, 13, 11, 9, 7, 15, 7, 12, 15, 9, 11, 7, 13, 12,
+	11, 13, 6, 7, 14, 9, 13, 15, 14, 8, 13, 6, 5, 12, 7, 5,
+	11, 12, 14, 15, 14, 15, 9, 8, 9, 14, 5, 6, 8, 6, 5, 12,
+	9, 15, 5, 11, 6, 8, 13, 12, 5, 12, 13, 14, 11, 8, 5, 6,
+}
+
+// same for the other parallel one
+var n_ = [80]uint{
+	5, 14, 7, 0, 9, 2, 11, 4, 13, 6, 15, 8, 1, 10, 3, 12,
+	6, 11, 3, 7, 0, 13, 5, 10, 14, 15, 8, 12, 4, 9, 1, 2,
+	15, 5, 1, 3, 7, 14, 6, 9, 11, 8, 12, 2, 10, 0, 4, 13,
+	8, 6, 4, 1, 3, 11, 15, 0, 5, 12, 2, 13, 9, 7, 10, 14,
+	12, 15, 10, 4, 1, 5, 8, 7, 6, 2, 13, 14, 0, 3, 9, 11,
+}
+
+var r_ = [80]uint{
+	8, 9, 9, 11, 13, 15, 15, 5, 7, 7, 8, 11, 14, 14, 12, 6,
+	9, 13, 15, 7, 12, 8, 9, 11, 7, 7, 12, 7, 6, 15, 13, 11,
+	9, 7, 15, 11, 8, 6, 6, 14, 12, 13, 5, 14, 13, 13, 7, 5,
+	15, 5, 8, 11, 14, 14, 6, 14, 6, 9, 12, 9, 12, 5, 15, 8,
+	8, 5, 12, 9, 12, 5, 14, 6, 8, 13, 6, 5, 15, 13, 11, 11,
+}
+
+var _k = uints.NewU32Array([]uint32{
+	0x50a28be6,
+	0x5a827999, 0x5c4dd124,
+	0x6ed9eba1, 0x6d703ef3,
+	0x8f1bbcdc, 0x7a6d76e9,
+	0xa953fd4e,
+})
+
+func Permute(uapi *uints.BinaryField[uints.U32], currentHash [5]uints.U32, p [64]uints.U8) (newHash [5]uints.U32) {
+	var x [16]uints.U32
+	var alpha, beta uints.U32
+	a, b, c, d, e := currentHash[0], currentHash[1], currentHash[2], currentHash[3], currentHash[4]
+	aa, bb, cc, dd, ee := a, b, c, d, e
+	for i := 0; i < 16; i++ {
+		x[i] = uapi.PackLSB(p[4*i], p[4*i+1], p[4*i+2], p[4*i+3])
+	}
+	// round 1
+	for i := 0; i < 16; i++ {
+		// alpha = a + (b ^ c ^ d) + x[_n[i]]
+		alpha = uapi.Add(
+			a,
+			uapi.Xor(b, c, d),
+			x[_n[i]],
+		)
+		// s := int(_r[i])
+		s := int(_r[i])
+		// alpha = bits.RotateLeft32(alpha, s) + e
+		alpha = uapi.Add(
+			uapi.Lrot(alpha, s),
+			e,
+		)
+		// beta = bits.RotateLeft32(c, 10)
+		beta = uapi.Lrot(c, 10)
+		// a, b, c, d, e = e, alpha, b, beta, d
+		a, b, c, d, e = e, alpha, b, beta, d
+
+		// parallel line
+		// alpha = aa + (bb ^ (cc | ^dd)) + x[n_[i]] + 0x50a28be6
+		alpha = uapi.Add(
+			aa,
+			uapi.Xor(bb,
+				uapi.Or(
+					cc,
+					uapi.Not(dd),
+				),
+			),
+			x[n_[i]],
+			_k[0],
+		)
+		// s = int(r_[i])
+		s = int(r_[i])
+		// alpha = bits.RotateLeft32(alpha, s) + ee
+		alpha = uapi.Add(
+			uapi.Lrot(alpha, s),
+			ee,
+		)
+		// beta = bits.RotateLeft32(cc, 10)
+		beta = uapi.Lrot(cc, 10)
+		// aa, bb, cc, dd, ee = ee, alpha, bb, beta, dd
+		aa, bb, cc, dd, ee = ee, alpha, bb, beta, dd
+	}
+
+	// round 2
+	for i := 16; i < 32; i++ {
```

Comment:
I think since the structure of the rounds is identical, it would be better to have a function `Round(a, b, c, d, e, f, x, k, s) -> a, b, c, d, e, f`, two arrays of function references for the `f` and two loops each iterating from 0 to 80.
I think it'd be shorter and read more like the paper.

Reply:
Refactored. The round function is a bit ugly as one K constant is 0. It is inefficient to add zero as we still perform range checking of the inputs etc, so added extra clauses. And also `f` is called with different indices depending on the lane.

But otherwise, as you said, looks a lot nicer.